randomize amounts and fees

danvergara opened the pull request:
Closes #427

This PR adds a function to randomize amounts within a range (+/- 5%) to break amount correlation attacks, but it's up to user to opt-in, just in case they prefer keep accounting under control.

This randomization is applied to amount spend in the coinswap, the transactions fees and the maker fees.

The ways this is kept optional is adding boolean flags to the cli commands.

![Screenshot From 2025-03-02 12-51-10](https://github.com/user-attachments/assets/23594da2-4eb0-45db-a21e-51f3a960da45)

![Screenshot From 2025-03-02 12-51-30](https://github.com/user-attachments/assets/87625046-5c37-4046-b553-fc469d9a32d2)

![Screenshot From 2025-03-02 12-52-15](https://github.com/user-attachments/assets/21625875-d0d7-4524-bc8d-10514b64137f)

mojoX911 commented on Cargo.toml:
```diff
@@ -29,6 +29,7 @@ log4rs = "1.3.0"
 flate2 = {version = "1.0.35", optional = true}
 tar = {version = "0.4.43", optional = true}
 minreq = { version = "2.12.0", features = ["https"] , optional = true}
+rand = "0.9.0"
```
rust bitcoin already comes with a rand crate. can we use that instead of a new dep?

danvergara replied:
I just removed that crate and used the one you suggested.

danvergara commented:
> Concept ACK. I like the optional design. At least it will be easier to get in first without breaking the tests.
> 
> Maybe for production we can turn this to default true, so users doesn't have to be aware about it.
> 
> Some nit fixes needed, and fmt fix. I will give it a full manual run soon.
> 
> Thanks for the work.

Thanks for the feedback, just addressed your comments and run the fmt on the code.

danvergara commented:
I just added `[allow(clippy::too_many_arguments)]` to the init method, so we can add a new argument.

danvergara commented:
Hey @mojoX911 can I get a review when you have the chance? If something is blocking this, just let me know.
Thanks in advance.

mojoX911 commented:
Hi @danvergara , Initial concept ACK and thanks for the work. Code wise it looks clean. 

But this work needs to be created on top #449 as that refactor will also impact how we can have feerate instead of fees inside funding transaction creation. Then we will need to randmize that feerate instead of how we are doing fees here. 

449 is up for grab and nobody assigned on it yet. So if you want you can take that up, and then we will rebase this PR on top of it.

danvergara commented:
Thanks for the quick response @mojoX911. Oh I see. Sure, I can take it up, as soon as I' done wit #443 (which is still in draft and hope to wrap it up this weekend).

NeoZ666 commented:
#449 is similar to #439 , I am working on it

danvergara commented:
If @mojoX911 is OK, I don't have any problem. Just I'd like to know why both issues exist.

mojoX911 commented:
> If @mojoX911 is OK, I don't have any problem. Just I'd like to know why both issues exist.

Both the issues are related. Just not linked properly. @NeoZ666 will take that up as he is working closely with the spend API. @danvergara feel free to join up with neo, as there will be more tasks in that direction soon, or feel free to take up any other issues/todos.

danvergara commented:
Sounds good @mojoX911. @NeoZ666 please keep us posted so I can rebase your changes properly. I will continue working on the Split/Merge PR in the meantime.

mojoX911 commented:
> Sounds good @mojoX911. @NeoZ666 please keep us posted so I can rebase your changes properly. I will continue working on the Split/Merge PR in the meantime.

The split/Merge approach will also require some more discussions and is dependent upon a few other protocol changes that we have been thinking about, like using a single transaction and many outputs to do outputs, randomizing inputs with both parties' contributions to further remove amount correlations, etc.

You can carry on the work, but expect redesign or refactors will be required.

NeoZ666 commented:
Yea. It's done and closed.

danvergara commented:
Hey @mojoX911 I just rebased this PR onto master, randomizing the fee rate.

mojoX911 commented:
Closing this as an approach will be done as per the split/merge project design.
